internal/adaptorsigs/adaptor.go

```diff
@@ -0,0 +1,509 @@
+// create a valid signature from the adaptor signature. Then, the valid
+// signature can be combined with the adaptor signature to reveal the hidden
+// value.
+func PublicKeyTweakedAdaptorSig(privKey *secp256k1.PrivateKey, hash []byte, T *secp256k1.JacobianPoint) (*AdaptorSignature, error) {
+	// The algorithm for producing an encrypted EC-Schnorr-DCRv0 is as follows:
+	// The deviations from the original algorithm are in step 5 and 6.
+	//
+	// G = curve generator
+	// n = curve order
+	// d = private key
+	// m = message
+	// r, s = signature
+	// t = hidden value
+	// T = t * G
+	//
+	// 1. Fail if m is not 32 bytes
+	// 2. Fail if d = 0 or d >= n
+	// 3. Use RFC6979 to generate a deterministic nonce k in [1, n-1]
+	//    parameterized by the private key, message being signed, extra data
+	//    that identifies the scheme, and an iteration count
+	// 4. R = kG
+	// 5. Repeat from step 3 if R + T is odd
+	// 6. r = (R + T).x (R.x is the x coordinate of the point R + T)
+	// 7. e = BLAKE-256(r || m) (Ensure r is padded to 32 bytes)
+	// 8. Repeat from step 3 (with iteration + 1) if e >= n
+	// 9. s = k - e*d mod n
+	// 10. Return (r, s)
+
+	// Step 1.
+	//
+	// Fail if m is not 32 bytes
+	if len(hash) != scalarSize {
+		str := fmt.Sprintf("wrong size for message hash (got %v, want %v)",
+			len(hash), scalarSize)
+		return nil, errors.New(str)
+	}
+
+	// Step 2.
+	//
+	// Fail if d = 0 or d >= n
+	privKeyScalar := &privKey.Key
+	if privKeyScalar.IsZero() {
+		str := "private key is zero"
+		return nil, errors.New(str)
+	}
+
+	var privKeyBytes [scalarSize]byte
+	privKeyScalar.PutBytes(&privKeyBytes)
+	defer zeroArray(&privKeyBytes)
+	for iteration := uint32(0); ; iteration++ {
+		// Step 3.
+		//
+		// Use RFC6979 to generate a deterministic nonce k in [1, n-1]
+		// parameterized by the private key, message being signed, extra data
+		// that identifies the scheme, and an iteration count
+		k := secp256k1.NonceRFC6979(privKeyBytes[:], hash, rfc6979ExtraDataV0[:],
+			nil, iteration)
+
+		// Steps 4-10.
+		sig, err := schnorrEncryptedSign(privKeyScalar, k, hash, T)
+		k.Zero()
+		if err != nil {
+			// Try again with a new nonce.
+			continue
+		}
+
+		return sig, nil
+	}
+}
+
+func parseSig(sig *schnorr.Signature) (r *secp256k1.FieldVal, s *secp256k1.ModNScalar) {
+	sigB := sig.Serialize()
+	r, s = new(secp256k1.FieldVal), new(secp256k1.ModNScalar)
+	r.SetBytes((*[32]byte)(sigB[0:32]))
+	s.SetBytes((*[32]byte)(sigB[32:64]))
+	return r, s
+}
+
+// PrivateKeyTweakedAdaptorSig creates a private key tweaked adaptor signature.
+// This is created by a party which knows the hidden value.
+func PrivateKeyTweakedAdaptorSig(sig *schnorr.Signature, pubKey *secp256k1.PublicKey, t *secp256k1.ModNScalar) *AdaptorSignature {
+	T := new(secp256k1.JacobianPoint)
+	secp256k1.ScalarBaseMultNonConst(t, T)
+	T.ToAffine()
+
+	r, s := parseSig(sig)
+	tweakedS := new(secp256k1.ModNScalar).Add2(s, t)
+
+	return &AdaptorSignature{
+		r: *r,
```

Comment:
This case with https://github.com/decred/dcrd/pull/3437 too:

```suggestion
	s := sig.S()
	tweakedS := new(secp256k1.ModNScalar).Add2(&s, t)

	return &AdaptorSignature{
		r: sig.R(),
```

Or the even more compact:

```Go
	s := sig.S()
	return &AdaptorSignature{
		r: sig.R(),
		s: *new(secp256k1.ModNScalar).Add2(&s, t),
		t: affinePoint{x: T.X, y: T.Y},
	}
```